profiling/utils.py

```diff
@@ -0,0 +1,364 @@
+        network = self.make_network(
+            model_dim, n_layers, is_distributed=False, config=config
+        )
+
+        hook_fn = functools.partial(hook_function_cpu, acc=self.cpu_acc)
+
+        self._hook_every_layer(network, hook_fn)
+
+        return network
+
+    def multi_gpu_no_hooks(self, model_dim, n_layers, config):
+        network = self.make_network(
+            model_dim, n_layers, is_distributed=True, config=config
+        )
+        return network
+
+    def multi_gpu_unity_hooks(self, model_dim, n_layers, config):
+        network = self.make_network(
+            model_dim, n_layers, is_distributed=True, config=config
+        )
+
+        hook_fn = hook_function_identity
+
+        self._hook_every_layer(network, hook_fn)
+
+        return network
+
+    def multi_gpu_cpu_hooks(self, model_dim, n_layers, config):
+        network = self.make_network(
+            model_dim, n_layers, is_distributed=True, config=config
+        )
+
+        # NOTE: This will not accumulate full tensors since we don't gather.
+        hook_fn = functools.partial(hook_function_cpu, acc=self.cpu_acc)
+
+        self._hook_every_layer(network, hook_fn)
+
+        return network
+
+    def multi_gpu_gpu_hooks(self, model_dim, n_layers, config):
+        network = self.make_network(
+            model_dim, n_layers, is_distributed=True, config=config
+        )
+
+        # NOTE: This will not accumulate full tensors since we don't gather.
+        hook_fn = functools.partial(hook_function_gpu, acc=self.cpu_acc)
+
+        self._hook_every_layer(network, hook_fn)
+
+        return network
+
+    def multi_gpu_cpu_hooks_with_gather_scatter(self, model_dim, n_layers, config):
+        network = self.make_network(
+            model_dim, n_layers, is_distributed=True, config=config
+        )
+        hook_fn = functools.partial(hook_function_cpu_gather_scatter, acc=self.cpu_acc)
+
+        self._hook_every_layer(network, hook_fn)
+
+        return network
+
+    def multi_gpu_flex_model(self, model_dim, n_layers, config):
+        base_network = self.make_network(
+            model_dim, n_layers, is_distributed=True, config=config
+        )
+
+        network = FlexModel(
+            base_network,
+            self.cpu_acc,
+            tensor_parallel_size=mpu.get_tensor_model_parallel_world_size(),
+            pipeline_parallel_size=mpu.get_pipeline_model_parallel_world_size(),
+            data_parallel_size=mpu.get_data_parallel_world_size(),
+        )
+
+        module_names_to_hook = set(
+            f"layers.{i}" for i in range(len(base_network.layers))
+        )
+        for n in module_names_to_hook:
+            network.register_forward_hook(
+                HookFunction(module_name=n, expected_shape=(None, model_dim),)
+            )
+        return network
+
+
+def init_megatron_dist(args):
+    """Initialize Megatron-LM parallel state."""
+    os.environ["NCCL_IB_DISABLE"] = "1"
+    initialize_distributed()
```

Comment:
there are lots of distributed backends being enabled at this point, specify this is torch distributed.